File: backend/app/models/knowledge.py
import uuid
from datetime import datetime
from typing import Optional, Any
from sqlmodel import SQLModel, Field, Column
from sqlalchemy import JSON, Text, UniqueConstraint


class KnowledgeSource(SQLModel, table=True):
    """The user's external knowledge base (Obsidian vault folder or Notion workspace).

    One source per user. Pulled periodically into vault_files/vault_chunks so
    chat + agents can search it via RAG.
    """
    __tablename__ = "knowledge_sources"
    __table_args__ = (UniqueConstraint("user_id", name="uq_knowledge_source_user"),)

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    user_id: uuid.UUID = Field(foreign_key="users.id", index=True, nullable=False)
    source_type: str = Field(nullable=False)  # "obsidian" | "notion"
    config: Optional[Any] = Field(default=None, sa_column=Column(JSON))  # obsidian: {"path": ...}
    enabled: bool = Field(default=True, nullable=False)
    sync_interval_minutes: int = Field(default=30, nullable=False)
    last_synced_at: Optional[datetime] = None
    last_status: Optional[str] = None  # "ok" | "error"
    last_error: Optional[str] = Field(default=None, sa_column=Column(Text))
    created_at: datetime = Field(default_factory=lambda: datetime.utcnow(), nullable=False)
    updated_at: datetime = Field(default_factory=lambda: datetime.utcnow(), nullable=False)

File: frontend/packages/shared/src/api/knowledge.ts
import { api } from './client'

export interface KnowledgeSourceState {
  configured: boolean
  folder_sync_available: boolean
  source_type?: 'obsidian' | 'notion'
  config?: { path?: string }
  enabled?: boolean
  sync_interval_minutes?: number
  last_synced_at?: string | null
  last_status?: 'ok' | 'error' | null
  last_error?: string | null
}

export interface KnowledgeSourceBody {
  source_type: 'obsidian' | 'notion'
  config?: { path?: string }
  sync_interval_minutes?: number
  enabled?: boolean
}

export const knowledgeApi = {
  get: () => api.get('/knowledge/source').then(r => r.data as KnowledgeSourceState),
  save: (body: KnowledgeSourceBody) => api.put('/knowledge/source', body).then(r => r.data as KnowledgeSourceState),
  syncNow: () => api.post('/knowledge/source/sync').then(r => r.data as { status: string }),
  remove: () => api.delete('/knowledge/source').then(r => r.data as { status: string }),
}
